### PR TITLE
new cmp_null lint (fixes #1184) WIP

### LINKED ISSUE #1184: Suggest using is_null instead of comparing to ptr::null()
I would use a lint for comparing a pointer to `std::ptr::null()` or `std::ptr::null_mut()`, instead of using the inherent `is_null` method on pointers. It would lint on

    if(handle == ptr::null_mut()) {
    }

so that it could be changed to

    if(handle.is_null()) {
    }

### PR DESCRIPTION
**Pull request by llogiq**
This crashes compiletest for me at the moment :anguished:

### CODE REVIEW

File: clippy_lints/src/lib.rs [hunk 1]
```diff
@@ -281,7 +281,6 @@ pub fn register_plugins(reg: &mut rustc_plugin::Registry) {
         methods::RESULT_UNWRAP_USED,
         methods::WRONG_PUB_SELF_CONVENTION,
         misc::USED_UNDERSCORE_BINDING,
-        misc_early::UNSEPARATED_LITERAL_SUFFIX,
```
Comment (mcarton) [inline clippy_lints/src/lib.rs]: ?

File: clippy_lints/src/ptr.rs [hunk 2]
```diff
@@ -5,16 +5,16 @@ use rustc::hir::map::NodeItem;
...
 ///
-/// **Why is this bad?** Requiring the argument to be of the specific size makes
-/// the function less useful for no benefit; slices in the form of `&[T]` or
-/// `&str` usually suffice and can be obtained from other types, too.
+/// **Why is this bad?** Requiring the argument to be of the specific size makes the function less
+/// useful for no benefit; slices in the form of `&[T]` or `&str` usually suffice and can be
+/// obtained from other types, too.
 ///
-/// **Known problems:** None.
+/// **Known problems:** None
```
Comment (mcarton) [inline clippy_lints/src/ptr.rs]: AFAIK all lints have a dot here.

File: clippy_lints/src/utils/ty.rs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+use rustc::ty::Ty;
+
+fn is_internally_mutable(ty: Ty) {
```
Comment (mcarton) [inline clippy_lints/src/utils/ty.rs]: This does not seem to be used anywhere.

File: clippy_lints/src/lib.rs (hunk 1 above)
Comment (llogiq) [inline clippy_lints/src/lib.rs]: No idea how I lost that one.

File: clippy_lints/src/utils/ty.rs (hunk 3 above)
Comment (llogiq) [inline clippy_lints/src/utils/ty.rs]: Oops, that was an early try to create a utility function that would help us find out if a function is pure (together with finding global state). I thought I'd stashed that. Egad, I'll remove it from the PR.

**Comment by mcarton** [comment]
@llogiq: tests fail because you forgot `./util/update_lints.py`, you need an `import std::ptr` in the test file, and there is a spelling mistake <tt>ptr::null_*n*ut</tt>.
Except for that, r+.

**Comment by llogiq** [comment]
Thanks, mcarton. There were also dogfood problems, which I fixed, too.

**Pull request merged**